### PR TITLE
8337632: AES-GCM Algorithm optimization for x86_64

### PR DESCRIPTION
**Pull request by smita-kamath**
Hi, 
I want to submit an AES-GCM algorithm optimization. This implementation is using AVX512/VAES Instructions. Additionally, it reduces PARALLEL_LEN from 7680 to 512 bytes. The performance numbers are as below. Kindly review the code. Thank you.

Benchmark | Datasize | BaseJDK (ops/s) | Patch(ops/s) | %Gain
-- | -- | -- | -- | --
full.AESGCMBench.decrypt | 512 | 2928259.197 | 3269964.387 | 11.67
full.AESGCMBench.decrypt | 1024 | 2494254.611 | 3010987.731 | 20.72
full.AESGCMBench.decrypt | 1500 | 1883453.546 | 1934915.846 | 2.73
full.AESGCMBench.decrypt | 2048 | 1825780.711 | 2452861.368 | 34.34
full.AESGCMBench.decrypt | 4096 | 1275108.345 | 1806329.066 | 41.66
full.AESGCMBench.decrypt | 8192 | 1033936.634 | 1196836.052 | 15.75
full.AESGCMBench.decrypt | 16384 | 681494.768 | 711630.498 | 4.42
full.AESGCMBench.decrypt | 32768 | 385026.017 | 395043.193 | 2.6
full.AESGCMBench.decrypt | 65536 | 207373.924 | 214723.588 | 3.54
  |   |   |   |  
full.AESGCMBench.encrypt | 512 | 2658008.476 | 2882496.94 | 8.45
full.AESGCMBench.encrypt | 1024 | 2283709.63 | 2589534.403 | 13.39
full.AESGCMBench.encrypt | 1500 | 1794993.519 | 1817669.531 | 1.26
full.AESGCMBench.encrypt | 2048 | 1745532.435 | 2191097.29 | 25.52
full.AESGCMBench.encrypt | 4096 | 1203301.174 | 1649593.953 | 37.08
full.AESGCMBench.encrypt | 8192 | 985174.988 | 1132407.54 | 14.94
full.AESGCMBench.encrypt | 16384 | 658980.441 | 684765.771 | 3.91
full.AESGCMBench.encrypt | 32768 | 373543.798 | 391518.837 | 4.81
full.AESGCMBench.encrypt | 65536 | 202532.315 | 205084.833 | 1.260301597

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337632](https://bugs.openjdk.org/browse/JDK-8337632): AES-GCM Algorithm optimization for x86_64 (**Enhancement** - P4)

### Reviewers
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**) Review applies to [ed10bcca](https://git.openjdk.org/jdk/pull/17515/files/ed10bccaa8215a47f1a2abdfc69bcbf7bb9e3237)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17515/head:pull/17515` \
`$ git checkout pull/17515`

Update a local copy of the PR: \
`$ git checkout pull/17515` \
`$ git pull https://git.openjdk.org/jdk.git pull/17515/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17515`

View PR using the GUI difftool: \
`$ git pr show -t 17515`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17515.diff">https://git.openjdk.org/jdk/pull/17515.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17515#issuecomment-2262099207)

### CODE REVIEW

**Comment by ascarpino** [comment]
That's a good performance increase for such a small code change.  I reviewed the simple java code change.  I'll let a hotspot reviewer handle the rest of the code.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:3480]: ghash16_encrypt_parallel16_avx512  needs to pass in CTL_BLOCKx, AAD_HASHx, ADDBE_4x4, ADDBE_1234, AAD_HASHx, SHUF_MASK just like we do in initial_blocks_16_avx512. Also the GL and GH needs to be passed in.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:3084]: Should this be:
__ cmpb(CTR_CHECK, (256 - 16)
__ jcc(Assembler::aboveEqual, blocks_overflow);

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:298]: It will be helpful to document the structure of subkeyHtbl (what is there in the 200 long entries).

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:2827]: You can define a macro for this repetitive sequence and pass varying argument to it for brevity.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:3433]: Could this handle be Assembler::less, in which case you can handle 256 byte length as well.

File: src/hotspot/cpu/x86/assembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/assembler_x86.cpp:8984]: As this is an evex instruction, we could call it evinserti64x2. Also the assert should be: assert(vector_len == AVX_512bit || VM_Version::supports_avx512vl(), "");

File: src/hotspot/cpu/x86/assembler_x86.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/assembler_x86.cpp:11057]: The assert should be: assert(vector_len == AVX_512bit || VM_Version::supports_avx512vl(), "");

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:2813]: This should be i > 0.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:203]: This is same as COUNTER_MASK_LINC0[] in this file. Could we reuse that instead?

**Comment by sviswa7** [comment]
Please also update copyright dates in stubGenerator_x86_64_aes.cpp and stubGenerator_x86_64_ghash.cpp files.

**Review by sviswa7 — APPROVED** [review]
Looks good to me now.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:3001]: Is this based on aggregate reduction method ? 
Can you please add some comments to narrate the reduction algorithm.

**Comment by jatin-bhateja** [comment]
Kindly extend to micro for small dataSizes (128,256 and 512) and keyLenght (128, 192)
https://github.com/openjdk/jdk/blob/master/test/micro/org/openjdk/bench/javax/crypto/full/AESGCMBench.java

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (smita-kamath) [reply src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:286]: I am using all XMM registers from 0-31 in the code, so wont be able to do this change.

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (smita-kamath) [reply src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:3001]: The reduction algorithm is mentioned in the paper - https://github.com/intel/intel-ipsec-mb/wiki/doc/advanced-encryption-standard-galois-counter-mode-optimized-ghash-function-technology-guide-1693300747.pdf

File: src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp:286]: Thanks for clarifying.

**Review by jatin-bhateja — APPROVED** [review]
Thanks @smita-kamath  for addressing comments.

**Comment by smita-kamath** [comment]
@ascarpino I have two approvals for this PR. Would it be possible for you to run this through your testing? Thanks a lot!

**Comment by ascarpino** [comment]
@smita-kamath ok.  I'll let you know when I have results.

**Comment by ascarpino** [comment]
All tests passed.  Thanks.

**Pull request closed without merging**
